javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -0,0 +1,153 @@
+    static int BRANCHING_BASE = 5;
+    static int branchingFactor()                   { return 1 << BRANCHING_BASE; }
+    static int firstDigit(int num, int depthShift) { return num >> depthShift; }
+    static int digit(int num, int depthShift)      { return lastDigit(firstDigit(num, depthShift)); }
+    static int lastDigit(int num)                  { return num & (-1 >>> -BRANCHING_BASE); }
+
+    private static final long serialVersionUID = 1L;
+
+    private static final BitMappedTrie<?> EMPTY = new BitMappedTrie<>(Arrays.empty(), 0, 0, 0);
+    @SuppressWarnings("unchecked")
+    static <T> BitMappedTrie<T> empty() { return (BitMappedTrie<T>) EMPTY; }
+
+    private final Object[] array;
+    private final int offset, length;
+    private final int depthShift;
+
+    BitMappedTrie(Object[] array, int offset, int length, int depthShift) {
+        this.array = array;
+        this.offset = offset;
+        this.length = length;
+        this.depthShift = depthShift;
+
+        assert length() <= treeSize(branchingFactor(), depthShift);
+    }
+
+    private static int treeSize(int branchCount, int depthShift) {
+        final int fullBranchSize = 1 << depthShift;
+        return branchCount * fullBranchSize;
+    }
+
+    BitMappedTrie<T> prependLeaf(T leading) {
+        final int newSize = length() + 1;
+        if (length() == 0) {
+            return new BitMappedTrie<>(asArray(leading), offset, newSize, depthShift);
+        } else {
+            Object[] array = this.array;
+            int depthShift = this.depthShift, offset = this.offset;
+            if (isFullLeft()) {
+                array = copyUpdate(Arrays.empty(), branchingFactor() - 1, array);
+                depthShift += BRANCHING_BASE;
+                offset = treeSize(branchingFactor() - 1, depthShift);
+            }
+
+            offset -= 1;
+            array = modifyLeaf(array, depthShift, offset, NODE_MODIFIER, LEAF_MODIFIER.apply(leading));
+            return new BitMappedTrie<>(array, offset, newSize, depthShift);
+        }
+    }
+    private boolean isFullLeft() { return offset == 0; }
+
+    BitMappedTrie<T> appendLeaf(T trailing) {
+        final int newSize = length() + 1;
+        if (length() == 0) {
+            return new BitMappedTrie<>(asArray(trailing), offset, newSize, depthShift);
+        } else {
+            Object[] array = this.array;
+            int depthShift = this.depthShift;
+            if (isFullRight(newSize)) {
+                array = asArray(array);
+                depthShift += BRANCHING_BASE;
+            }
+
+            array = modifyLeaf(array, depthShift, offset + length(), NODE_MODIFIER, LEAF_MODIFIER.apply(trailing));
+            return new BitMappedTrie<>(array, offset, newSize, depthShift);
+        }
+    }
+    private boolean isFullRight(int newSize) { return (offset + newSize) > treeSize(branchingFactor(), depthShift); }
+
+    private static final NodeModifier NODE_MODIFIER = (o, i) -> copy(o, i + 1);
```

Comment:
You could give the field `NODE_MODIFIER` a name that reflects *what it does* as opposed to *what it is*. You would do the same if, instead of a field, you would have a no-arg method that returned you the  `NodeModifier` instance.

I understand that this `NodeModifier` copies the given array and makes sure that `i + 1` becomes a valid index (increasing the array size if needed). One suggestion would be to name it `COPY_NODE` (ignoring the part about the array size, as I'm not sure if that's really important).